### ISSUE TITLE
New feture: open CircuitBreaker for a time based on the decorated calls response

### ISSUE THREAD

**Issue by walec51** [resilience4j/resilience4j#1387]
Problem:

We have a client class that uses resiliance4j decorators. The underlining API, which the client calls, can ban our IP for some time for multiple reasons. The underlining API in its error response returns the duration of the given ban and expects that we will stop any further communication during it. Not doing so could result in an even longer ban.

Solution:

This is an ideal use case for a circuit breaker as the bans duration can exceed the period of any rate limiter that is also applied. However resilience4j circuit breaker lacks configuration options and some implementation mechanisms to handle this. Therefore I propose to add the following API:

```
CircuitBreakerConfig.from(...)
                // ...
                .forceOpenOnResult(Either<Throwable, T> result -> {
                    if (result.getLeft() instanceof SomeSpecificException) {
                        SomeSpecificException e = (SomeSpecificException) result.getLeft();
                        return ForceOpenChecksResult.neededUntil(e.getBannedUntil()); // Instant
                    }
                    return ForceOpenChecksResult.notNeeded();
                })
                .build()
``` 

`ForceOpenChecksResult` could also have a creator method called `ForceOpenChecksResult.needed()` that would force open the circuit breaker for the time configured in `waitDurationInOpenState` attribute.

I'm willing to implement this feature if it will be accepted. We need this feature in this project: https://github.com/knowm/XChange where we already use resilience4j for rate limiting and retries.

**Comment by RobWin** [comment]
I like the idea that you can handle results and based on the result you can force a transition to open. 
Comparable to `RateLimiterConfig.drainPermissionsOnResult` predicate, we could add a `CircuitBreakerConfig.forceOpenOnResult()` predicate.

**Comment by walec51** [comment]
@RobWin to handle my use case I would need something more then a predicate - I'm thinking more of an `Function<ForceOpenChecksResult, Either<Throwable, T>>` as shown in the example

**Comment by walec51** [comment]
in `ForceOpenChecksResult` one would pass if a force open is needed and optionally until which point in time should it be opened

**Comment by RobWin** [comment]
Ok, 
some ideas:

1) We have to add a `private void transitionToOpenState(long waitDurationInMillis)` to the `CircuitBreakerStateMachine` and add another constructor to `CircuitBreakerStateMachine.OpenState` so that you can use a fixed waitDurationInMillis and not use the `waitIntervalFunctionInOpenState` from the CircuitBreakerConfig.

2) We should not call the methd `forceOpenOnResult`, because `FORCED_OPEN` is a state of its own. Maybe we should better call it `transitionToOpenStateOnResult`. 

3) I would like to avoid creating an `ForceOpenChecksResult` object on every call.  We should rename the class  to `TransitionToOpenCheckResult`.

`TransitionToOpenCheckResult.noTransition()` should return a singleton.
- `TransitionToOpenCheckResult.transitionToOpen()` means that the CircuitBreaker transitions to OPEN, but uses the waitDurationInMillis from the Config.
- `TransitionToOpenCheckResult.transitionToOpen(waitDurationInMillis)` means that the CircuitBreaker transitions to OPEN, but uses the the given waitDurationInMillis.
- `TransitionToOpenCheckResult.transitionToOpen(instant)` means that the CircuitBreaker transitions to OPEN but calculate the waitDurationInMillis based on the given Instant.

We could also a `CircuitBreakerResultUtils` comparable to the `io.github.resilience4j.core,ResultUtils which simplifies the usage:

```
.transitionToOpenStateOnResult(
                callResult -> CircuitBreakerResultUtils.isFailedAndThrown(callsResult, SomeSpecificException .class, (ex) -> TransitionToOpenCheckResult.transitionToOpen(ex.getBannedUntil()))
```

or

```
.transitionToOpenStateOnResult(
                callResult -> CircuitBreakerResultUtils.isSuccessfulAndReturned(callsResult, Response.class, (response) -> TransitionToOpenCheckResult.transitionToOpen(response.getBannedUntil()))
```

**Comment by RobWin** [comment]
Hi @walec51 
Are you still interested to implement this feature?

**Comment by walec51** [comment]
@RobWin yes, I have read all your suggestions and I agree with them, I just didn't have the time to start working on this - maybe I'll find some on this weekend

**Comment by AhsanSheraz** [comment]
+1 — server-hinted wait times are the missing piece in most resilience setups; fixed
`waitDurationInOpenState` fights the server instead of cooperating with it.

For anyone arriving here from the Python ecosystem looking for the same capability: I maintain
pyresilience (a Resilience4j-inspired Python library), and as of 0.4.0 the retry layer can take
its delay directly from the response — `retry_after_delay()` parses `Retry-After`
(delta-seconds and HTTP-date) and falls back to exponential backoff when absent. Would love to
see this land in Resilience4j itself for the JVM side.
